sandbox/modules/identity/src/main/java/org/opensearch/identity/authz/OpenSearchPermission.java

```diff
@@ -10,27 +10,169 @@
-import java.util.Objects;
+import java.util.HashSet;
+import java.util.Set;
+import java.util.Collections;
+import java.util.ArrayList;
 import java.util.stream.Collectors;
+import java.util.stream.Stream;
 
 public class OpenSearchPermission implements Permission {
 
-    private String actionName;
+    private List<String> resourcePatterns;
+    private final String PERMISSION_DELIMITER = "\\.";
 
-    private List<String> indexPatterns;
+    public final String RESOURCE_DELIMITER = ",";
+
+    private String permissionString;
+    private String[] permissionSegments;
+
+    private String action;
+    private String permissionType;
 
     public OpenSearchPermission(String permission) {
-        Objects.requireNonNull(permission);
-        String[] permissionParts = permission.split("\\|");
-        // turns an action like indices:data/read/search into indices:data:read:search to leverage shiro's WildcardPermission
-        this.actionName = permissionParts[0].replace("/", ":");
-        if (permissionParts.length > 1) {
-            this.indexPatterns = Arrays.stream(permissionParts[1].split(",")).map(String::trim).collect(Collectors.toList());
+
+        this.permissionString = permission;
+        try {
+            this.permissionSegments = permissionString.split(PERMISSION_DELIMITER);
+            this.permissionType = permissionSegments[0];
+            this.action = permissionSegments[1];
+        } catch (IndexOutOfBoundsException ex) {
+            throw new PermissionFactory.InvalidPermissionException(
+                "All permissions must contain a permission type and" + " action delimited by a " + PERMISSION_DELIMITER + "."
+            );
+        }
+        // Handle two legacy index permissions that are really cluster permissions
+        if (this.action.equalsIgnoreCase("admin/template/") || this.action.equalsIgnoreCase("admin/index_template/")) {
+            this.resourcePatterns = Arrays.asList("*");
+        }
+        if (this.permissionSegments.length == 3) {
+            String resourceString = permissionSegments[2];
+            this.resourcePatterns = Arrays.asList(resourceString.split(RESOURCE_DELIMITER));
+
+        }
+    }
+
+    public String getPermissionType() {
+        return this.permissionType;
+    }
+
+    public String getAction() {
+        return this.action;
+    }
+
+    public List<String> getResource() {
+        return this.resourcePatterns;
+    }
+
+    public String getPermissionString() {
+        return this.permissionString;
+    }
+
+    /**
+     * Compare the current permission's permission type to another permission's permission type
+     */
+    public boolean permissionTypesMatch(OpenSearchPermission permission) {
+
+        if (this.permissionType.equalsIgnoreCase(permission.permissionType)) {
+            return true;
         }
+        return false;
+    }
+
+    /**
+     * Resolves complex resource patterns with '-' signs to the corresponding resource patterns
```

Comment:
Nice to see an implementation, but just want to voice reservations about adding negative permissions.

IMO I'd rather be able to decompose an incoming request like `GET /index-6*,-index-666` so we don't check for permission of `index-666`, but avoid letting declaration the permissions in that way.

Reply:
Yeah, I know in the past you had mentioned being hesitant on this. I just wanted to put this here and show it working so we can check it off the todos. It is something that had been discussed with permission granting and I would rather have it ready to go should we decide it is worthwhile then not have it. I understand your thinking though and think that I would be happy with either.